### PR TITLE
feat(operator): delete, enable, disable, backups

### PR DESCRIPTION
**Pull request by Dav-14**
FEAT: ENG-103

# Operator

- Add flag to disable the finalizer
- Add finalizer on stack object

Service drop idempotency:
-  PG: Drop PG DATA with `IF EXISTS`
-  ELK: DELETE /_query/<INDEX_STACKS> is already idempotent

Brokers idempotency:
- NATS: Check subject existance then delete if exist

# Agent
- According with `membership/feat/disable_stack`
- Merge Disabled on Stack (ExistingStackMessage)

### CODE REVIEW

**Comment by Dav-14** [comment]
1. `garden deploy`
2. `kubectl apply -f ./garden/example-v1beta3.yaml`
3. DELETE de la stack a la main
4. `kubectl port-forward svc/minio 9001:9001`
5. Enjoy S3 backuping

File: components/agent/internal/grpc/client.go [hunk 1]
```diff
@@ -246,13 +247,51 @@ func (client *client) Start(ctx context.Context) error {
...
-						StackId: stack.Name,
-						Status:  status,
+						ClusterName: stack.Name,
+						Status:      status,
 					},
 				},
 			}); err != nil {
 				sharedlogging.FromContext(ctx).Errorf("Unable to send stack status to server: %s", err)
 			}
+
```
Comment (gfyrag) [inline components/agent/internal/grpc/client.go:257]: The ExistingStach message is intended for the agent, not for membership.

File: components/agent/internal/grpc/client.go [hunk 2]
```diff
@@ -246,13 +247,51 @@ func (client *client) Start(ctx context.Context) error {
...
+							return timestamppb.New(stack.DeletionTimestamp.Time)
+						}(),
+					},
+				},
+			}); err != nil {
+				sharedlogging.FromContext(ctx).Errorf("Unable to send ExistingStack to membership: %s", err)
+			}
+
+			sharedlogging.FromContext(ctx).Infof("Stack[STATUS] %s updated cluster side", stack.Name)
```
Comment (gfyrag) [inline components/agent/internal/grpc/client.go:278]: The status is already sent earlier

File: components/operator/internal/controllers/stack/delete/brokers.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	"fmt"
+	"reflect"
+
+	"github.com/formancehq/operator/apis/stack/v1beta3"
+	"github.com/formancehq/operator/internal/controllers/stack/storage/nats"
+	"github.com/go-logr/logr"
+)
+
+var (
+	natsClientId = "12"
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/delete/brokers.go:14]: use a meaningful name

File: components/operator/internal/controllers/stack/delete/brokers.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+)
+var (
+	ErrCast = errors.New("cannot cast interface to string")
+)
+
+func DeleteByBrokers(c *v1beta3.Configuration, stackName string, subjectService []string, logger logr.Logger) error {
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/delete/brokers.go:20]: No need to use reflection as you know the structure of the configuration

File: components/operator/internal/controllers/stack/storage/nats/client.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot connect to nats-core")
+	}
+
+	return conn, nil
+}
+
+func ExistSubject(conn *nats.Conn, subject string, logger logr.Logger) (bool, error) {
+
+	js, _ := conn.JetStream()
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/storage/nats/client.go:25]: handler errors

File: components/operator/internal/controllers/stack/storage/nats/client.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+	_, err := js.StreamNameBySubject(subject)
+	if err != nil {
+		if err.Error() == "nats: no stream matches subject" {
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/storage/nats/client.go:29]: avoid compare error by string. Check if nats implements some kind of IsErrXXX()

File: components/operator/internal/controllers/stack/storage/nats/client.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+		return false, errors.Wrap(err, "cannot get stream info")
+	}
+
+	return true, nil
+
+}
+
+func DeleteSubject(conn *nats.Conn, subject string) error {
+
+	js, _ := conn.JetStream()
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/storage/nats/client.go:41]: handle error

File: components/operator/internal/controllers/stack/storage/pg/function.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	_, err := db.Exec("DROP DATABASE IF EXISTS " + fmt.Sprintf("\"%s-%s\"", stackName, serviceName))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func OpenClient(config v1beta3.PostgresConfig) (*sql.DB, error) {
+	debug := true
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/storage/pg/function.go:45]: make configurable

**Review by gfyrag — CHANGES_REQUESTED** [review]
need to extract psql logic

File: components/agent/internal/grpc/client.go [hunk 9]
```diff
@@ -272,37 +273,41 @@ func (client *client) Start(ctx context.Context) error {
...
 				sharedlogging.FromContext(ctx).Infof("Stack %s deleted", msg.DeletedStack.ClusterName)
 			case *generated.Order_DisabledStack:
+				sharedlogging.FromContext(ctx).Infof("Incomming DISABLING: %s", msg.DisabledStack.ClusterName)
+
 				existingStack, err := client.k8sClient.Get(ctx, msg.DisabledStack.ClusterName, metav1.GetOptions{})
 				if err != nil {
 					if controllererrors.IsNotFound(err) {
 						sharedlogging.FromContext(ctx).Infof("Cannot disable not existing stack: %s", msg.DisabledStack.ClusterName)
-						continue
+						return err
```
Comment (gfyrag) [inline components/agent/internal/grpc/client.go:310]: You don't want to return there

File: components/agent/internal/grpc/client.go [hunk 10]
```diff
@@ -314,11 +319,12 @@ func (client *client) Start(ctx context.Context) error {
 				}
 				sharedlogging.FromContext(ctx).Infof("Stack %s disabled", msg.DisabledStack.ClusterName)
 			case *generated.Order_EnabledStack:
+				sharedlogging.FromContext(ctx).Infof("Incomming ENABLING: %s", msg.EnabledStack.ClusterName)
 				existingStack, err := client.k8sClient.Get(ctx, msg.EnabledStack.ClusterName, metav1.GetOptions{})
 				if err != nil {
 					if controllererrors.IsNotFound(err) {
 						sharedlogging.FromContext(ctx).Infof("Cannot enable not existing stack: %s", msg.EnabledStack.ClusterName)
-						continue
+						return err
```
Comment (gfyrag) [inline components/agent/internal/grpc/client.go:327]: You don't want to return there

File: components/agent/internal/grpc/client.go [hunk 11]
```diff
@@ -272,37 +273,41 @@ func (client *client) Start(ctx context.Context) error {
...
+
+						return err
 					}
+					sharedlogging.FromContext(ctx).Errorf("Get Stack cluster side: %s", err)
+					continue
+				}
+
+				if err := client.k8sClient.Delete(ctx, msg.DeletedStack.ClusterName); err != nil {
+					sharedlogging.FromContext(ctx).Errorf("Deleting cluster side: %s", err)
+					continue
```
Comment (paul-nicolas) [inline components/agent/internal/grpc/client.go:300]: No error if failed to be deleted ?

File: components/operator/Dockerfile [hunk 12]
```diff
@@ -13,8 +13,9 @@ RUN --mount=type=cache,id=gobuild,target=/root/.cache/go-build \
 
 # Use distroless as minimal base image to package the manager binary
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot as release
+FROM alpine:3.18.2 as release
 LABEL org.opencontainers.image.source=https://github.com/formancehq/operator
+RUN apk add --no-cache postgresql15-client
```
Comment (gfyrag) [inline components/operator/Dockerfile]: Restore Dockerfile

File: components/agent/internal/grpc/client.go [hunk 13]
```diff
@@ -272,37 +273,41 @@ func (client *client) Start(ctx context.Context) error {
...
-									Status:  generated.StackStatus_Deleted,
-								},
-							},
-						}); err != nil {
-							sharedlogging.FromContext(ctx).Errorf("Unable to send stack status to server: %s", err)
-							continue
-						}
+						sharedlogging.FromContext(ctx).Infof("Cannot delete not existing stack: %s", msg.DeletedStack.ClusterName)
+
+						return err
```
Comment (paul-nicolas) [inline components/agent/internal/grpc/client.go:292]: same comment as @gfyrag below, should we return here ?

File: components/operator/build.Dockerfile [hunk 14]
```diff
@@ -1,4 +1,9 @@
 FROM ghcr.io/formancehq/base:22.04
```
Comment (gfyrag) [inline components/operator/build.Dockerfile:1]: Restore Dockerfile

File: components/operator/apis/stack/v1beta3/postgres.go [hunk 15]
```diff
@@ -28,6 +28,9 @@ type PostgresConfig struct {
 	Host     string `json:"host"`
 	Username string `json:"username"`
 	Password string `json:"password"`
+
+	// +optional
+	Debug bool `json:"debug"`
```
Comment (gfyrag) [inline components/operator/apis/stack/v1beta3/postgres.go:33]: Why this property?

File: components/operator/internal/controllers/stack/delete/brokers.go (hunk 4 above)
Comment (gfyrag) [reply components/operator/internal/controllers/stack/delete/brokers.go:20]: Same comment again :D
You can simply use configuration.Spec.Broker.Nats != nil (something like this)

File: components/operator/internal/controllers/stack/delete/brokers.go [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+		return err
+	}
+	defer client.Close()
+
+	for _, service := range subjectService {
+		stackServiceSubject := fmt.Sprintf("%s-%s", stackName, service)
+
+		exist, err := nats.ExistSubject(client, stackServiceSubject, logger)
+		if err != nil {
+			logger.Error(err, "NATS: subject existancy check")
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/delete/brokers.go:50]: Avoir writing too many logs if you return the error.
Prefer return the error, you can wrap it if you want, and display it at top level.

File: components/operator/internal/controllers/stack/stack_finalizer.go [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	finalizerName = "backup-delete"
+)
+
+var (
+	subjects = []string{"ledger", "payments"}
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/stack_finalizer.go:21]: Don't hardcode list of subject.
The registered module have a "NeedTopic" property.
This inform the core to create the topic.
You can retrieve it and dynamically retrieve the list of subjects

File: components/operator/go.mod [hunk 18]
```diff
@@ -28,6 +28,8 @@ require (
 	sigs.k8s.io/controller-runtime v0.14.4
 )
 
+require github.com/jmespath/go-jmespath v0.4.0 // indirect
```
Comment (paul-nicolas) [inline components/operator/go.mod:31]: why not with the others ?

File: components/operator/internal/controllers/stack/storage/es/function.go [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+		logger.Error(err, "ES client error")
+		return err
+	}
+
+	var (
+		buf bytes.Buffer
+		res Response
+	)
+
+	query := map[string]interface{}{
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/storage/es/function.go:28]: I think you can replace all this stuff by some method provided by the ES library

File: components/operator/internal/controllers/stack/storage/nats/client.go [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+
+	conn, err := nats.Connect(natsConfig.URL, options...)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot connect to nats-core")
+	}
+
+	return conn, nil
+}
+
+func ExistSubject(conn *nats.Conn, subject string, logger logr.Logger) (bool, error) {
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/storage/nats/client.go:23]: Nats doesn"t have methods to pass a context?

File: components/operator/internal/controllers/stack/delete/brokers.go [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+func DeleteByBrokers(c *v1beta3.Configuration, stackName string, subjectService []string, logger logr.Logger) error {
+	values := reflect.ValueOf(c.Spec.Broker)
+	for i := 0; i < values.NumField(); i++ {
+		if rv := reflect.ValueOf(values.Field(i)); !rv.IsValid() || rv.IsNil() {
+			continue
+		}
+		switch t := values.Field(i).Interface().(type) {
+		case v1beta3.NatsConfig:
```
Comment (paul-nicolas) [inline components/operator/internal/controllers/stack/delete/brokers.go:27]: We handle only NATS ? Should we not handle kafka as well if clients want to use kafka ? (I don't know, just asking the question)

File: components/operator/internal/controllers/stack/stack_finalizer.go [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+	if err := delete.DeleteByService(conf, stack.Name, log); err != nil {
+		return &ctrl.Result{
+			Requeue:      true,
+			RequeueAfter: time.Second,
+		}, err
+	}
+
+	log.Info("start deleting brokers subjects " + stack.Name)
+	if err := delete.DeleteByBrokers(conf, stack.Name, subjects, log); err != nil {
+		return nil, err
```
Comment (paul-nicolas) [inline components/operator/internal/controllers/stack/stack_finalizer.go:96]: No requeue here ?

File: components/operator/internal/controllers/stack/delete/brokers.go (hunk 21 above)
Comment (Dav-14) [reply components/operator/internal/controllers/stack/delete/brokers.go:27]: Hum, that's a question, at begining i was asked to only do NATS, I can also do it for Kafka !

File: components/operator/apis/stack/v1beta3/postgres.go (hunk 15 above)
Comment (Dav-14) [reply components/operator/apis/stack/v1beta3/postgres.go:33]: With this 
```
func OpenClient(config v1beta3.PostgresConfig) (*sql.DB, error) {
	return OpenSQLDB(ConnectionOptions{
		DatabaseSourceName: config.DSN(),
		Debug:              config.Debug,
		Trace:              config.Debug,
		Writer:             os.Stdout,
		MaxIdleConns:       20,
		ConnMaxIdleTime:    time.Minute,
		MaxOpenConns:       20,
	})
}
```

File: components/operator/internal/controllers/stack/stack_reconciler.go [hunk 23]
```diff
@@ -114,6 +114,23 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	}
+
+	res, err := r.HandleFinalizer(ctx, log, stack, conf, req)
+	if err != nil {
+		return *res, err
+	}
+
+	if res != nil {
+		return *res, err
+	}
```
Comment (Dav-14) [inline components/operator/internal/controllers/stack/stack_reconciler.go:132]:
```suggestion
	if err != nil {
		return ctrl.Result{}, err
	}

	if res != nil {
		return *res, err
	}
```

File: components/operator/internal/controllers/stack/storage/es/function.go (hunk 19 above)
Comment (Dav-14) [reply components/operator/internal/controllers/stack/storage/es/function.go:28]:
```suggestion
func DropESIndex(config *v1beta3.ElasticSearchConfig, logger logr.Logger, stackName string) error {
	client, err := NewElasticSearchClient(config)
	if err != nil {
		logger.Error(err, "ES client error")
		return err
	}

	response, err := client.Delete(stacksIndex, stackName)

	if err != nil {
		logger.Error(err, "ELK: Error during index drop")
		return err
	}

	defer response.Body.Close()
	if response.StatusCode > 300 {
		logger.Error(fmt.Errorf("ELK: Status code: %d", response.StatusCode), "ELK: Error during index drop")
		return fmt.Errorf("ELK: Status code: %d", response.StatusCode)
	}

	logger.Info(fmt.Sprintf("ELK Status code: %d", response.StatusCode))

	var res Response
	if err := json.NewDecoder(response.Body).Decode(&res); err != nil {
		logger.Error(err, "ELK: Error decoding es body")
		return err
	}

	if response.StatusCode == 200 {
		logger.Info("ES Index Dropped")
	}

	return nil
}
```suggestion

```
```

Comment (Dav-14) [reply components/operator/internal/controllers/stack/storage/es/function.go]:
```
func DropESIndex(config *v1beta3.ElasticSearchConfig, logger logr.Logger, stackName string) error {
	client, err := NewElasticSearchClient(config)
	if err != nil {
		logger.Error(err, "ES client error")
		return err
	}

	response, err := client.Delete(stacksIndex, stackName)

	if err != nil {
		return err
	}

	if response.IsError() {
		return fmt.Errorf("ELK status: %d", response.StatusCode)
	}

	return nil
}
```

File: components/operator/internal/controllers/stack/stack_reconciler.go [hunk 24]
```diff
@@ -114,6 +114,19 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+	if err := r.client.Get(ctx, types.NamespacedName{
+		Namespace: "",
+		Name:      stack.Spec.Seed,
+	}, conf); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	res, err := r.HandleFinalizer(ctx, log, stack, conf, req)
+	if res != nil || err != nil {
+		return *res, err
```
[the remark this answers is not in the document]
Comment (Dav-14) [reply components/operator/internal/controllers/stack/stack_reconciler.go:131]:
```suggestion
		    if err != nil {
        return ctrl.Result{}, err
    }

    if res != nil {
        return *res, err
    }
```

File: components/agent/internal/grpc/client.go [hunk 25]
```diff
@@ -272,32 +273,41 @@ func (client *client) Start(ctx context.Context) error {
...
 				if _, err := client.k8sClient.Update(ctx, newStack); err != nil {
 					sharedlogging.FromContext(ctx).Errorf("Updating stack cluster side: %s", err)
 					continue
 				}
-				sharedlogging.FromContext(ctx).Infof("Stack %s updated", newStack.Name)
+
+				sharedlogging.FromContext(ctx).Infof("Stack %s updated cluster side", newStack.Name)
 
 			case *generated.Order_DeletedStack:
+
```
Comment (gfyrag) [inline components/agent/internal/grpc/client.go:285]: Why getting the stack before deleting it?

File: components/operator/internal/controllers/stack/stack_reconciler.go [hunk 26]
```diff
@@ -114,6 +114,33 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+		ctx,
+		r.client,
+		log,
+		r.configuration,
+		stack,
+		conf,
+		&stackv1beta3.Versions{},
+	)
+
+	res, err := stackFinalizer.HandleFinalizer(ctx, log, stack, conf, req)
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/stack_reconciler.go:135]: Your finalizer should indicate the object was deleted to allow this function to stop the process.

File: components/operator/internal/modules/module.go [hunk 27]
```diff
@@ -131,10 +131,10 @@ func (sd *StackDeployer) HandleStack(ctx Context, deployer *ResourceDeployer) (b
 		}
 	}
 
-	if finalizer, ok := podDeployer.(interface {
+	if StackFinalizer, ok := podDeployer.(interface {
```
Comment (gfyrag) [inline components/operator/internal/modules/module.go:134]: you fails your rename with your IDE ;)

File: components/operator/internal/storage/es/client.go [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+	Deleted  int                      `json:"deleted"`
+	Failures []map[string]interface{} `json:"failures"`
+	Took     int                      `json:"took"`
+	TimedOut bool                     `json:"timed_out"`
+}
+
+func NewElasticSearchClient(config *v1beta3.ElasticSearchConfig) (*opensearch.Client, error) {
+	httpTransport := http.DefaultTransport
+	httpTransport.(*http.Transport).TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: true,
```
Comment (gfyrag) [inline components/operator/internal/storage/es/client.go:25]: There is a config in v1beta3.ElasticsearcgConfig for this

File: components/operator/internal/storage/es/client.go [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	opensearchConfig := opensearch.Config{
+		Addresses: []string{config.Scheme + "://" + config.Host + ":" + strconv.FormatUint(uint64(config.Port), 10)},
```
Comment (gfyrag) [inline components/operator/internal/storage/es/client.go:29]: There is a function Endpoint() on the config to compute this value

File: components/operator/internal/controllers/stack/stack_finalizer.go [hunk 30]
```diff
@@ -0,0 +1,299 @@
...
+
+func (f *StackFinalizer) deleleNatsSubjects(config *v1beta3.NatsConfig, stackName string, subjectService []string, logger logr.Logger) error {
+	client, err := nats.NewClient(config, natsClientId)
+	if err != nil {
+		logger.Error(err, "NATS: client")
+		return err
+	}
+	defer client.Close()
+
+	jsCtx, err := client.JetStream()
```
Comment (gfyrag) [inline components/operator/internal/controllers/stack/stack_finalizer.go:198]: Maybe this client has to be closed?

Comment (Dav-14) [reply components/operator/internal/controllers/stack/stack_finalizer.go:196]: This is the context of Nats Jetstream, seems not being related to connection as i don't see any `.Close` or any method in the sdk. The client is already close one line above

File: components/operator/internal/controllers/stack/stack_reconciler.go (hunk 26 above)
Comment (Dav-14) [reply components/operator/internal/controllers/stack/stack_reconciler.go:135]: There is watch/polling. As the operator has to be idempotent, I just try to get the ressource and see if i get a 404

Comment (Dav-14) [reply components/operator/internal/controllers/stack/stack_reconciler.go]: Now just a boolean when the process suceed

Comment (Dav-14) [reply components/operator/internal/controllers/stack/stack_reconciler.go]: Else if there is an error, we requeue

File: components/operator/internal/modules/service.go [hunk 31]
```diff
@@ -243,6 +243,57 @@ func (h ConfigHandle) GetMountPath() string {
...
+	// Environment is the environment the stack is deployed to: staging,
+	// production, sandbox, etc.
+	Environment   string
+	Stack         *stackv1beta3.Stack
+	Configuration *stackv1beta3.Configuration
+	Versions      *stackv1beta3.Versions
+}
+type ContextFn func(*Context) *Context
+
+func NewContext(
```
Comment (gfyrag) [inline components/operator/internal/modules/service.go:259]: Is this used somewhere?

File: components/operator/internal/storage/es/client.go [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+			},
+		},
+	}
+
+	if err := json.NewEncoder(&buf).Encode(query); err != nil {
+		logger.Error(err, "ELK: Error during json encoding")
+
+		return err
+	}
+	body := bytes.NewReader(buf.Bytes())
```
Comment (gfyrag) [inline components/operator/internal/storage/es/client.go:57]: Forward context everywhere you can

File: components/operator/garden/example-v1beta3.yaml [hunk 33]
```diff
@@ -139,7 +140,7 @@ spec:
 apiVersion: stack.formance.com/v1beta3
 kind: Stack
 metadata:
-  name: stack1
```
Comment (gfyrag) [inline components/operator/garden/example-v1beta3.yaml:142]: Restore original config please ;)

**Review by gfyrag — APPROVED** [review]

**Pull request merged**